drivers/dai/nxp/esai/esai.c

```diff
@@ -0,0 +1,754 @@
+				      uint32_t bclk_rate, bool variable_hclk,
+				      bool allow_bclk_configuration,
+				      struct esai_transceiver_config *cfg)
+{
+	uint32_t hclk_div_ratio, bclk_div_ratio;
+
+	/* sanity checks */
+	if (!cfg) {
+		LOG_ERR("got NULL clock configuration");
+		return -EINVAL;
+	}
+
+	if (!extal_rate || !hclk_rate || !bclk_rate) {
+		LOG_ERR("got NULL clock rate");
+		return -EINVAL;
+	}
+
+	if (hclk_rate > extal_rate) {
+		LOG_ERR("HCLK rate cannot be higher than EXTAL rate");
+		return -EINVAL;
+	}
+
+	if (bclk_rate > extal_rate) {
+		LOG_ERR("BCLK rate cannot be higher than EXTAL rate");
+		return -EINVAL;
+	}
+
+	if (DIV_ROUND_UP(extal_rate, bclk_rate) > 2 * 8 * 256 * 16) {
+		LOG_ERR("BCLK rate %u cannot be obtained from EXTAL rate %u",
+			bclk_rate, extal_rate);
+		return -EINVAL;
+	}
+
+	/* TODO: add explanation */
+	if (DIV_ROUND_UP(extal_rate / 2, bclk_rate) == 1) {
+		LOG_ERR("HCLK prescaler bypass with divider bypass is not supported");
+		return -EINVAL;
+	}
+
+	hclk_div_ratio = 1;
+	bclk_div_ratio = 1;
+
+	/* check if HCLK is in (EXTAL_RATE / 2, EXTAL_RATE). If so,
+	 * return an error as any rates from this interval cannot be obtained.
+	 */
+	if (hclk_rate > extal_rate / 2 && hclk_rate < extal_rate) {
+		LOG_ERR("HCLK rate cannot be higher than EXTAL's rate divided by 2");
+		return -EINVAL;
+	}
+
+	/* compute HCLK configuration - only required if HCLK pad output is used */
+	if (!variable_hclk) {
+		if (extal_rate == hclk_rate) {
+			/* HCLK rate from pad is the same as EXTAL rate */
+			cfg->hclk_bypass = true;
+		} else {
+			/* EXTAL is automatically divided by 2 */
+			extal_rate /= 2;
+
+			/* compute prescaler divide ratio w/ prescaler bypass */
+			hclk_div_ratio = DIV_ROUND_UP(extal_rate, hclk_rate);
+
+			if (hclk_div_ratio > 256) {
+				/* can't obtain HCLK w/o prescaler */
+				cfg->hclk_prescaler_en = true;
+
+				extal_rate /= 8;
+
+				/* recompute ratio w/ prescaler */
+				hclk_div_ratio = DIV_ROUND_UP(extal_rate, hclk_rate);
+
+				if (hclk_div_ratio > 256) {
+					LOG_ERR("cannot obtain HCLK rate %u from EXTAL rate %u",
+						hclk_rate, extal_rate);
+					return -EINVAL;
+				}
+			}
+		}
+	}
+
+	cfg->hclk_div_ratio = hclk_div_ratio;
+
+	if (!allow_bclk_configuration) {
+		return 0;
+	}
+
+	extal_rate = DIV_ROUND_UP(extal_rate, hclk_div_ratio);
+
+	/* compute BCLK configuration */
+	if (variable_hclk || cfg->hclk_bypass) {
+		/* attempt to find a configuration that satisfies BCLK's rate */
+		extal_rate /= 2;
+
+		hclk_div_ratio = DIV_ROUND_UP(extal_rate, bclk_rate);
+
+		/* check if prescaler is required */
+		if (hclk_div_ratio > 256 * 16) {
+			extal_rate /= 8;
+			cfg->hclk_prescaler_en = true;
+			hclk_div_ratio = DIV_ROUND_UP(extal_rate, bclk_rate);
+		}
+
+		/* check if we really need to loop through TPM div ratios */
+		if (hclk_div_ratio < 256) {
+			cfg->bclk_div_ratio = 1;
+			cfg->hclk_div_ratio = hclk_div_ratio;
+			return 0;
+		}
+
+		for (int i = 1; i < 256; i++) {
```

Comment:
I see you use a lot of 2, 8, 16, 256. 
I suggest to have a macro for each one with an explicit name (why these numbers are used for ESAI clocks).

Reply:
I've added a comment regarding the clock generation and division near the function that takes care of the clocking. Hopefully this'll clear things up. Ideally I'd like to avoid the usage of such macros as I think it reduces code visibility in some places. Let me know if the comment is enough.